Sources/BSON/Document.swift

```diff
@@ -1,77 +1,156 @@
-    public init() { fatalError("Unimplemented") }
+    /// Initializes a new, empty `BSONDocument`.
+    public init() {
+        self.keySet = Set()
+        self._buffer = BSON_ALLOCATOR.buffer(capacity: 5)
+        self._buffer.writeInteger(5, endianness: .little, as: Int32.self)
+        self._buffer.writeBytes([0])
+    }
 
     /**
-     * Initializes a new `Document` from the provided BSON data.
+     * Initializes a new `BSONDocument` from the provided BSON data.
      *
      * - Throws:
      *   - `InvalidArgumentError` if the data passed is invalid BSON
      *
      * - SeeAlso: http://bsonspec.org/
      */
-    public init(fromBSON bson: Data) throws { fatalError("Unimplemented") }
+    public init(fromBSON bson: Data) throws {
+        var buffer = BSON_ALLOCATOR.buffer(capacity: bson.count)
+        buffer.writeBytes(bson)
+        try BSONDocument.validate(buffer)
+        self = BSONDocument(fromUnsafeBSON: buffer)
+    }
+
+    internal init(fromUnsafeBSON bson: Data) {
+        // trust the incoming format
+        var buffer = BSON_ALLOCATOR.buffer(capacity: bson.count)
+        buffer.writeBytes(bson)
+        self = BSONDocument(fromUnsafeBSON: buffer)
+    }
 
     /**
-     * Initializes a new `Document` from the provided BSON data.
+     * Initializes a new `BSONDocument` from the provided BSON data.
      *
      * - Throws:
      *   - `InvalidArgumentError` if the data passed is invalid BSON
      *
      * - SeeAlso: http://bsonspec.org/
      */
-    public init(fromBSON bson: ByteBuffer) throws { fatalError("Unimplemented") }
+    public init(fromBSON bson: ByteBuffer) throws {
+        try BSONDocument.validate(bson)
+        self = BSONDocument(fromUnsafeBSON: bson)
+    }
+
+    internal init(fromUnsafeBSON bson: ByteBuffer) {
+        self.keySet = Set()
+        self._buffer = bson
+        for (key, _) in self {
+            self.keySet.insert(key)
+        }
+    }
 
-    /// The keys in this `Document`.
-    public var keys: [String] { fatalError("Unimplemented") }
+    /// The keys in this `BSONDocument`.
+    public var keys: [String] { self.map { key, _ in key } }
 
-    /// The values in this `Document`.
-    public var values: [BSON] { fatalError("Unimplemented") }
+    /// The values in this `BSONDocument`.
+    public var values: [BSON] { self.map { _, val in val } }
 
     /// The number of (key, value) pairs stored at the top level of this document.
-    public var count: Int { fatalError("Unimplemented") }
+    public var count: Int { self.keySet.count }
 
     /// A copy of the `ByteBuffer` backing this document, containing raw BSON data. As `ByteBuffer`s implement
     /// copy-on-write, this copy will share byte storage with this document until either the document or the returned
     /// buffer is mutated.
-    public var buffer: ByteBuffer { fatalError("Unimplemented") }
+    public var buffer: ByteBuffer { ByteBuffer(self._buffer.readableBytesView) }
```

Comment:
Given that this property is making an explicit copy, I don't think the CoW behavior is actually happening. To enable CoW, just return the buffer itself (e.g. `self._buffer`).

Reply:
I changed it but just as an experiment I verified there's no change, I guess the initializer is smart enough not to touch the underlying storage unless necessary.